datasource/src/main/scala/quasar/physical/mongo/Tunnel.scala

```diff
@@ -0,0 +1,137 @@
+    new ConnectionString(s)
+  }
+
+  def mkJSch[F[_]: Sync]: F[JSch] = Sync[F].delay { new JSch() }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Null"))
+  def mkSession[F[_]: Sync](jsch: JSch, cfg: TunnelConfig): F[Session] = cfg.pass match {
+    case None =>
+      Sync[F].delay { jsch.getSession(cfg.user, cfg.host, cfg.port) }
+    case Some(cred) => cred match {
+      case Password(password) => for {
+        s <- Sync[F].delay { jsch.getSession(cfg.user, cfg.host, cfg.port) }
+        _ <- Sync[F].delay { s.setPassword(password) }
+      } yield s
+      case Identity(prv, passphrase) => for {
+        _ <- Sync[F].delay { jsch.addIdentity(SessionName, prv.getBytes("UTF-8"), null, passphrase map (_.getBytes("UTF-8")) getOrElse null) }
+        s <- Sync[F].delay { jsch.getSession(cfg.user, cfg.host, cfg.port) }
+      } yield s
+    }
+  }
+
+  def userInfo(cfg: TunnelConfig): UserInfo = new UserInfo {
+    override def getPassword(): String = cfg.pass match {
+      case None => ""
+      case Some(pass) => pass match {
+        case Identity(_, _) => ""
+        case Password(p) => p
+      }
+    }
+    override def getPassphrase(): String = cfg.pass match {
+      case None => ""
+      case Some(pass) => pass match {
+        case Identity(_, p) => p getOrElse ""
+        case Password(_) => ""
+      }
+    }
+    override def promptYesNo(s: String): Boolean = true
+    override def promptPassphrase(s: String): Boolean = true
+    override def promptPassword(s: String): Boolean = true
+    override def showMessage(s: String): Unit = ()
+  }
+
+  def setUserInfo[F[_]: Sync](s: Session, u: UserInfo): F[Unit] = Sync[F].delay {
+    s.setUserInfo(u)
+  }
+
+  def serverAddress[F[_]: Sync](settings: ClusterSettings): F[ServerAddress] = {
+    val hosts = settings.getHosts
+    if (hosts.size < 1) Sync[F].raiseError(new Throwable("Server addresses were not provided"))
+    else Sync[F].delay { hosts.get(0) }
+  }
+
+  final case class Tunnel(port: Int)
+
+  def openTunnel[F[_]: Sync](session: Session, address: ServerAddress): F[Tunnel] = Sync[F].delay {
+    session.connect()
+  } productR Sync[F].delay {
+    session.setPortForwardingL(0, address.getHost, address.getPort)
+  } map (Tunnel(_))
+
+  def closeTunnel[F[_]: Sync](session: Session, tunnel: Tunnel): F[Unit] = Sync[F].delay {
+    session.delPortForwardingL(tunnel.port)
+  } productR Sync[F].delay {
+    session.disconnect()
+  }
+
+  def tunneledSettings[F[_]: Sync](settings: ClusterSettings, tunnel: Tunnel): F[ClusterSettings] = Sync[F].delay {
+    ClusterSettings.builder(settings).hosts(List(new ServerAddress("localhost", tunnel.port)).asJava).build
+  }
+
+  def viaTunnel[F[_]: Sync: ContextShift](
+      config: TunnelConfig,
+      settings: ClusterSettings,
+      blockingPool: BlockingContext)
+      : F[Disposable[F, ClusterSettings]] = for {
+    jsch <- mkJSch
+    sess <- mkSession(jsch, config)
```

Comment:
Wondering if `mkSession` should also be done on the the `blockingPool`, it does some `synchronized` stuff here https://github.com/is/jsch/blob/e7d60f7f24dd490465eb1cf02a98cb9ce69ddfc5/src/main/java/com/jcraft/jsch/JSch.java#L169, though it's not a problem usually I guess

Reply:
Not worth to use `blockingPool` just for this, but it may be good to do just everything from `mkJSch` to `tunneledSettings` in a single `evalOn`. Again, just wondering, feel free to leave as is.